### ISSUE TITLE
Cancel notifications by target

### ISSUE THREAD

**Issue by atymic** [thomasjohnkane/snooze#11]
I think we should add a way to cancel all future notifications for a target. For example, if they unsub we don't want to continue sending.

Possible implementation:
Add a `target_type` column to the migration, and method to cancel outstanding notifications for a target.

**Comment by atymic** [comment]
Maybe requires us to store the primary key of the notifiable? Otherwise, we'd have to unserialize every single notification.

**Comment by thomasjohnkane** [comment]
I stored the notifiable primary key in the original version of this (back when it was only users *facepalm*). We could maybe just search the serialized data in mysql?

Seems cleaner to store the primary key, but we still would need to know what model we're looking at right? User vs Instructor for example. Seems like we would need "target_type" and "notifiable_id" unless we just try to search the serialized data.

**Comment by atymic** [comment]
Yep, we're currently storing `type` which is the notification type.

Maybe something like this?
```
$table->increments('id');

$table->string('target_id');
$table->string('target_type');
$table->text('target');
$table->string('notification_type');
$table->text('notification');

$table->datetime('send_at');

$table->datetime('sent_at')->nullable();
$table->datetime('rescheduled_at')->nullable();
$table->datetime('cancelled_at')->nullable();

$table->timestamps();
```

**Comment by atymic** [comment]
Since we haven't released a stable version, we should make sure we get the table structure right before we go to v1

I think i'm the only one using it in prod :P

**Comment by thomasjohnkane** [comment]
Haha yeah - I'm not using it yet, so your the OG user.

That structure looks good. The structure doesn't really matter to the end user, right?

**Comment by thomasjohnkane** [comment]
well I guess it would if they use the `create` helper...

**Comment by atymic** [comment]
No, but ideally we won't be changing it too much after v1 releases (because it's messy & we'd need multiple migrations)

**Comment by atymic** [comment]
> well I guess it would if they use the create helper...

That abstracts the logic anyway, so it's not an issue.

**Comment by thomasjohnkane** [comment]
Oh right - you just need to pass the notifiable for this and we sort the rest.

I'm good with this change and I agree now is the time to do it. Can you think of anything else we might want to ship with?

**Comment by atymic** [comment]
I'll rack my brain and see if I can think of anything :)

**Comment by thomasjohnkane** [comment]
Merged in PR#20
